core/google/api/core/gapic_v1/method.py

```diff
@@ -99,33 +133,22 @@ def __init__(self, target, retry, timeout, metadata):
 
     def __call__(self, *args, **kwargs):
         """Invoke the low-level RPC with retry, timeout, and metadata."""
-        # Note: Due to Python 2 lacking keyword-only arguments we use this
-        # pattern to extract the retry and timeout params.
-        retry_ = kwargs.pop('retry', None)
-        if retry_ is None:
-            retry_ = self._retry
-
-        timeout_ = kwargs.pop('timeout', None)
-        if timeout_ is None:
-            timeout_ = self._timeout
-
-        # If timeout is specified as a number instead of a Timeout instance,
-        # convert it to a ConstantTimeout.
-        if timeout_ is not False and isinstance(timeout_, (int, float)):
-            timeout_ = timeout.ConstantTimeout(timeout_)
-
-        # If timeout is the default and the default timeout is exponential and
-        # a non-default retry is specified, make sure the timeout's deadline
-        # matches the retry's. This handles the case where the user leaves
-        # the timeout default but specifies a lower deadline via the retry.
-        if (timeout_ is self._timeout
-                and retry_ is not False
-                and retry_ is not self._retry
-                and isinstance(timeout_, timeout.ExponentialTimeout)):
-            timeout_ = timeout_.with_deadline(retry_._deadline)
+        # Note: Due to Python 2 lacking keyword-only arguments we kwargs to
+        # to extract the retry and timeout params.
+        timeout_ = _determine_timeout(
+            self._timeout,
+            kwargs.pop('timeout', None),
+            # Use the invocation-specified retry only for this, as we only
+            # want to adjust the timeout deadline if the *user* specified
+            # a different retry.
+            kwargs.get('retry', None))
+
+        retry = kwargs.pop('retry', None)
```

Comment:
1. Why not pop before setting `timeout_` and then just pass the popped value into `_determine_timeout`?
1. Can we avoid the `timeout` vs. `timeout_` issue just by using `google.api.core.timeout` or `core.timeout`